### PR TITLE
SW-5293 Implemented GET one module API

### PR DESCRIPTION
**Pull request by tommylau523**

### CODE REVIEW

**Comment by tommylau523** [comment]
* **#1936** <a href="https://app.graphite.dev/github/pr/terraware/terraware-server/1936?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1901** <a href="https://app.graphite.dev/github/pr/terraware/terraware-server/1901?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1895** <a href="https://app.graphite.dev/github/pr/terraware/terraware-server/1895?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1894** <a href="https://app.graphite.dev/github/pr/terraware/terraware-server/1894?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1878** <a href="https://app.graphite.dev/github/pr/terraware/terraware-server/1878?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#1891** <a href="https://app.graphite.dev/github/pr/terraware/terraware-server/1891?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1884** <a href="https://app.graphite.dev/github/pr/terraware/terraware-server/1884?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @tommylau523 and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: src/main/kotlin/com/terraformation/backend/accelerator/api/ProjectModulesController.kt [hunk 1]
```diff
@@ -21,20 +21,36 @@ import org.springframework.web.bind.annotation.RequestMapping
...
+
+  @ApiResponse200
+  @ApiResponse404
+  @GetMapping("/{moduleId}")
+  @Operation(summary = "Gets one module for a project.")
+  fun getModule(
+      @PathVariable("projectId") projectId: ProjectId,
+      @PathVariable moduleId: ModuleId,
+  ): GetProjectModuleResponsePayload {
+    val models = moduleStore.fetchModulesForProject(projectId)
```
Comment (tommylau523) [inline src/main/kotlin/com/terraformation/backend/accelerator/api/ProjectModulesController.kt]: Cannot use fetch one here because we do need the list of modules for ordering. 

There is probably a SQL query that can do something like `join select row_num() from .... ` but it is pretty complex and will require changes to our data models.

Comment (tommylau523) [reply src/main/kotlin/com/terraformation/backend/accelerator/api/ProjectModulesController.kt]: if the SQL method is preferred we can track in as a separate task.

Comment (sgrimm) [reply src/main/kotlin/com/terraformation/backend/accelerator/api/ProjectModulesController.kt]: I guess you mean you need the index? What's that used for? Could the start date (which is unique and sortable) serve the same purpose?

Comment (tommylau523) [reply src/main/kotlin/com/terraformation/backend/accelerator/api/ProjectModulesController.kt]: The start date is unique and sortable. I think this has something to do with how our client selector is set up. @nickgraz can you give some context?

Comment (nickgraz) [reply src/main/kotlin/com/terraformation/backend/accelerator/api/ProjectModulesController.kt]: We need to know what the modules "title" is. For example "name: Kick-off, title: Module 1". When we load a module, we need to know what the title is, but the data isn't stored in a field it seems, it is derived from the position in the list instead.

Comment (sgrimm) [reply src/main/kotlin/com/terraformation/backend/accelerator/api/ProjectModulesController.kt]: I'd kind of be inclined to say that the user-facing title of a module should be an explicit value we specify in the spreadsheet, rather than something we calculate based on date ordering.

I could easily see us, say, wanting "Module 3A" and "Module 3B" at some point.

Comment (tommylau523) [reply src/main/kotlin/com/terraformation/backend/accelerator/api/ProjectModulesController.kt]: Let me add the title in cohorts_modules table then.

**Review by sgrimm — CHANGES_REQUESTED** [review]
It seems like the admin UI changes got folded into this PR, along with the schema change to support user-supplied titles; can you split the changes up so this one is just the read path? It'll be easier to review if it's not changing multiple things at once.

**Comment by tommylau523** [comment]
> It seems like the admin UI changes got folded into this PR, along with the schema change to support user-supplied titles; can you split the changes up so this one is just the read path? It'll be easier to review if it's not changing multiple things at once.
> 
> 

Done! Most of the changes are to do with the title column, which I had split into a separate PR.

File: src/main/kotlin/com/terraformation/backend/accelerator/api/ProjectModulesController.kt [hunk 2]
```diff
@@ -21,20 +21,32 @@ import org.springframework.web.bind.annotation.RequestMapping
...
-    return GetProjectModulesResponsePayload(models.map { ProjectModule(it) })
+    return GetProjectModulesResponsePayload(models.map { model -> ProjectModule(model) })
+  }
+
+  @ApiResponse200
+  @ApiResponse404
+  @GetMapping("/{moduleId}")
+  @Operation(summary = "Gets one module for a project.")
+  fun getModule(
+      @PathVariable("projectId") projectId: ProjectId,
```
Comment (sgrimm) [inline src/main/kotlin/com/terraformation/backend/accelerator/api/ProjectModulesController.kt]: No need for the name argument on the annotation here since the default is the parameter name. Doesn't hurt anything, just looks inconsistent since it's not present on the other parameter's annotation.

I know the existing code is inconsistent about specifying names; part of it is historical reasons (the server started off not using Spring Boot, and the other framework we used needed the explicit names) and part of it is just me being sloppy. At some point we should go clean it up across the code base.

File: src/main/kotlin/com/terraformation/backend/accelerator/db/CohortStore.kt [hunk 3]
```diff
@@ -118,6 +116,31 @@ class CohortStore(
...
+                .set(START_DATE, it.startDate)
+                .set(END_DATE, it.endDate)
+                .onDuplicateKeyUpdate()
+                .set(TITLE, it.title)
+                .set(START_DATE, it.startDate)
+                .set(END_DATE, it.endDate)
+                .execute()
+          }
+        }
+      }
```
Comment (sgrimm) [inline src/main/kotlin/com/terraformation/backend/accelerator/db/CohortStore.kt]: This feels like it belongs in another PR -- it has nothing to do with the GET endpoint, right?

File: src/main/kotlin/com/terraformation/backend/admin/AdminCohortsController.kt [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+import org.springframework.web.bind.annotation.RequestMapping
+import org.springframework.web.bind.annotation.RequestParam
+import org.springframework.web.servlet.mvc.support.RedirectAttributes
+
+@Controller
+@RequestMapping("/admin")
+@RequireGlobalRole(
+    [GlobalRole.SuperAdmin, GlobalRole.AcceleratorAdmin, GlobalRole.TFExpert, GlobalRole.ReadOnly])
+@Validated
+class AdminCohortsController(
```
Comment (sgrimm) [inline src/main/kotlin/com/terraformation/backend/admin/AdminCohortsController.kt]: This is both here and in #1884 but it seems like it should only be in #1884.

File: src/main/resources/db/migration/0250/V266__CohortModuleTitle.sql [hunk 5]
```diff
@@ -0,0 +1,11 @@
...
+ALTER TABLE accelerator.cohort_modules ADD COLUMN title TEXT;
+UPDATE accelerator.cohort_modules t
+SET title = CONCAT('Module ', s.row_num::TEXT)
+    FROM (
+        SELECT cohort_id, start_date, ROW_NUMBER() OVER (PARTITION BY cohort_id ORDER BY start_date) AS row_num
+        FROM accelerator.cohort_modules
+    ) AS s
+WHERE t.cohort_id = s.cohort_id AND t.start_date = s.start_date;
+
+ALTER TABLE accelerator.cohort_modules ALTER COLUMN title SET NOT NULL;
```
Comment (sgrimm) [inline src/main/resources/db/migration/0250/V266__CohortModuleTitle.sql]: These two migrations seem to have leaked in from other PRs.

**Review by sgrimm — CHANGES_REQUESTED** [review]
Seems like the PRs have gotten a bit jumbled.

**Comment by tommylau523** [comment]
Sorry looks like I accidentally did a force push without stacking. Should be resolved now!

**Review by sgrimm — APPROVED** [review]
Ah, much smaller now! Looks good.

**Comment by tommylau523** [comment]
### Merge activity

* **Apr 29, 3:07 PM EDT**: [Graphite](https://app.graphite.dev/github/pr/terraware/terraware-server/1878) rebased this pull request as part of a merge.
* **Apr 29, 3:15 PM EDT**: @tommylau523 merged this pull request with [Graphite](https://app.graphite.dev/github/pr/terraware/terraware-server/1878).
